File: cpp/56.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

struct Interval {
  int start;
  int end;
  Interval() : start(0), end(0) {}
  Interval(int s, int e) : start(s), end(e) {}
};

bool cmp(const Interval &a, const Interval &b){
  return a.start < b.start || (a.start == b.start && a.end < b.end);
}

class Solution {
public:
  vector<Interval> merge(vector<Interval> &intervals) {
    sort(begin(intervals), end(intervals), cmp);
    vector<Interval> ans;
    for(auto &v : intervals){
      if(ans.empty() || ans.back().end < v.start)
        ans.push_back(v);
      else
        ans.back().end = max(ans.back().end, v.end);
    }
    return ans;
  }
};

int main(){
  auto sol = Solution();
}

File: python/56.py
class Interval(object):
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e


class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        intervals.sort(key=lambda it: (it.start, it.end))
        res = []
        for it in intervals:
            if not res or res[-1].end < it.start:
                res.append(it)
            elif res[-1].end < it.end:
                res[-1].end = it.end
        return res
